### ISSUE TITLE
FR: Support git+ssh:// protocol

### ISSUE THREAD

**Issue by bendem** [containers/buildah#4032]
Currently, it is possible to build an image from a git repository by prefixing the context argument with `git://` or postfixing with `.git`. Sadly, this uses the raw git protocol without authentification. Most users want the https protocol or git over ssh (git+ssh).

Since the clone is delegated to the git command, you need to specify either `ssh://` or `git+ssh://` to use that protocol. Neither of those prefixes is recognized by buildah.

I have added the relevant lines in the code (types.go, `strings.HasPrefix(url, "git+ssh://")` in the two places required), but I'm not able to comprehend the code further. It looks from my testing like the cloning happens in an isolated namespace which doesn't have access either to ~/.ssh or to the `SSH_AUTH_SOCK`, whether `--ssh` is used or not.

Would love to hear what others think about this and whether it would be possible to allow building images from ssh repositories.

```
> buildah build --debug 'git+ssh://git@github.com/containers/buildah.git'
DEBU[0000] cloning "git+ssh://git@github.com/containers/buildah.git" to "/var/tmp/buildah1505040640"
error prepping temporary context directory: cloning "git+ssh://git@github.com/containers/buildah.git" to "/var/tmp/buildah1505040640":
Cloning into '/var/tmp/buildah1505040640'...
Bad owner or permissions on /etc/ssh/ssh_config.d/05-redhat.conf
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
: exit status 128
DEBU[0000] exit status 128
```

```
Version:         1.27.0-dev
Go Version:      go1.17.7
Image Spec:      1.0.2-dev
Runtime Spec:    1.0.2-dev
CNI Spec:        1.0.0
libcni Version:  v1.1.1
image Version:   5.21.2-dev
Git Commit:      fb28761c
Built:           Fri Jun  3 14:31:44 2022
OS/Arch:         linux/amd64
BuildPlatform:   linux/amd64
```

**Comment by rhatdan** [comment]
@nalind @flouthoc PTAL

**Comment by flouthoc** [comment]
@bendem Does your patch works when you run with `sudo` or `root` user ? I think issues is with running `git` from unshared session.

**Comment by bendem** [comment]
Indeed, it does work when using root. The git clone works, It's failing due to iptables errors, but I'm guessing that's an entirely different problem.

**Comment by bendem** [comment]
For reference, docker checks URLs a bit different and thus doesn't support `git+ssh` either. URL has to start with `git@`, `git://` or being an http url ending with `.git` + ref/subfolder.

https://github.com/docker/cli/blob/8dd94d382409b7384d210a2ed84b52f0d4a12995/vendor/github.com/docker/docker/builder/remotecontext/urlutil/urlutil.go#L79

**Comment by rhatdan** [comment]
Please open a PR with your patch.

**Comment by flouthoc** [comment]
I think issue is more like that user inside `unshare` session cannot access the global ssh config and altering file on host just for making it work for buildah  rootless looks weird to me, I'll try playing with this patch @bendem does this feature works on docker ? But even if it works its not a surprise since everything is accessed from `root` user there which is same when buildah is invoked from `root` user.

**Comment by bendem** [comment]
It only works in docker if your git server uses git as the connecting user since you the only URL format that will use git+ssh is 
`git@$server[#ref][:directory]`.

i.e. `docker build -t buildah 'git@github.com:containers/buildah.git#main:/contrib/docker'`

ref: https://github.com/docker/cli/blob/master/vendor/github.com/docker/docker/builder/remotecontext/urlutil/urlutil.go#L78

**Comment by bendem** [comment]
Opened https://github.com/docker/cli/issues/3702

**Comment by bendem** [comment]
Docker side mentions me that `+` in urls is deprecated (no source) and that they support (doesn't actually work) `ssh://`, which makes more sense in my opinion too.

**Comment by rhatdan** [comment]
@bendem @flouthoc What is going on with this issue?

**Comment by bendem** [comment]
Nothing, I don't have the knowledge to implement this, but it's still an open feature request.

**Comment by rhatdan** [comment]
@flouthoc WDYT?

**Comment by flouthoc** [comment]
> I think issue is more like that user inside `unshare` session cannot access the global ssh config and altering file on host just for making it work for buildah rootless looks weird to me, I'll try playing with this patch @bendem does this feature works on docker ? But even if it works its not a surprise since everything is accessed from `root` user there which is same when buildah is invoked from `root` user.

This is blocked because of this reason but lets wait for reply on docker issue first and see how it goes there: https://github.com/docker/buildx/issues/1209

**Comment by rhatdan** [comment]
Looks like not much is happening here.

**Comment by rhatdan** [comment]
@flouthoc PTAL
